`Bug` [Samsung_HVAC]: Build Error on 2026.2.0

### The problem

See attached build log.

### ESPHome Version

2026.2.0

### Samsung HVAC Component Version

v2026.01.001

### Home Assistant Version

2026.2.2

### AC Type

NON NASA

### Indoor AC Unit Model

N/A

### Outdoor AC Unit Model

N/A

### ESP Device Model

M5stack Atom Lite

### Connection Points

N/A

### ESPHome Logs

INFO ESPHome 2026.2.0
INFO Reading configuration /config/bedroomsamsungac.yaml...
INFO Updating https://github.com/omerfaruk-aran/esphome_samsung_hvac_bus.git@main
INFO Generating C++ source...
INFO Setting CONFIG_LWIP_MAX_SOCKETS to 14 (registered: api=4, captive_portal=4, mdns=2, ota=1, web_server=3)
INFO Compiling app... Build path: /data/build/bedroomsamsungac
Processing bedroomsamsungac (board: m5stack-atom; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.37/platform-espressif32.zip)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - contrib-piohome @ 3.4.4 
 - framework-espidf @ 3.50502.0 (5.5.2) 
 - tool-cmake @ 4.0.3 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.1.2 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-xtensa-esp-elf @ 14.2.0+20251107
Reading CMake configuration...
Dependency Graph
|-- noise-c @ 0.1.10
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/samsung_ac/debug_mqtt.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/samsung_ac/protocol_nasa.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/samsung_ac/protocol_non_nasa.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/samsung_ac/samsung_ac.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/samsung_ac/samsung_ac_device.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/samsung_ac/util.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/select/select.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/select/select_call.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/select/select_traits.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/sensirion_common/i2c_sensirion.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/sensor/automation.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/sensor/filter.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/sensor/sensor.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/sha256/sha256.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/sht3xd/sht3xd.cpp.o
Compiling .pioenvs/bedroomsamsungac/src/esphome/components/socket/bsd_sockets_impl.cpp.o
src/esphome/components/samsung_ac/debug_mqtt.cpp:8:10: fatal error: mqtt_client.h: No such file or directory

***************************************************************************
* Looking for mqtt_client.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:mqtt_client.h"
* Web  > https://registry.platformio.org/search?q=header:%1B%5Bm%1B%5BKmqtt_client.h
*
***************************************************************************

    8 | #include <mqtt_client.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
*** [.pioenvs/bedroomsamsungac/src/esphome/components/samsung_ac/debug_mqtt.cpp.o] Error 1
========================== [FAILED] Took 9.30 seconds ==========================

### Samsung AC Logs

_No response_

### YAML Configuration

```yaml
# This example provides a basic configuration and description based on the M5STACK components
# For further configuration you should visit the ESPHome docs https://esphome.io
# Every line starting with a # can be removed.

## BASIC CONFIGURATION
esphome:
  name: bedroomsamsungac
  friendly_name: Bedroom Samsung AC

#Test to fix mcp4725 dropout
  on_boot:
    priority: -100
    then:
      - delay: 2s


# Select chipset and board configuration - in this case M5STACK
esp32:
  board: m5stack-atom
  framework:
    type: esp-idf



# [Optional] - Create web interface where you can control the AC from your web browser (without Home Assistant)
web_server:
  port: 80

# Enable logging
logger:
  level: DEBUG
  # logs:
  #   component: ERROR # Remove the "Your component takes to long to respond warning"

# Enable Home Assistant API
api:

# Enable OTA firmware update, you will be able to upload the firmware over WiFi without connecting the device by cables
ota:
  - platform: esphome
    password: !secret ota_pwd

# Setup your WiFi password
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  power_save_mode: light
  fast_connect: true

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap: {}


captive_portal:


i2c:
  sda: GPIO26
  scl: GPIO32
  scan: True
  id: bus_a
  frequency: 100kHz
  timeout: 10ms

sensor:
  #Sensor from Home assistant
  - platform: homeassistant
    id: ha_temperature
    entity_id: sensor.bedroom_ac_samsung_workaround
    internal: true  # optional, prevents exposing to HA again
  #Sensor inside Air Conditioner
  - platform: sht3xd
    temperature:
      name: "Air Conditioner Temperature"
      id: ac_temperature
      filters:
        - offset: 0.6
    humidity:
      name: "Air Conditioner Humidity"
      id: ac_humidity
      filters:
        - offset: 0
    address: 0x44
    update_interval: 30s

text_sensor:
  - platform: template
    name: "AC Temperature Source"
    id: temp_source
    update_interval: 10s


output:
  - platform: mcp4725
    id: dac_output
    address: 0x60
    # no gain; we will set level 0.0..1.0 where 1.0 == Vdac_supply


interval:
  - interval: 10s
    then:
      - lambda: |-
          static int last_source = -1;  // tracks if source changed
          int current_source = 0;       // 0=HA, 1=SHT30, 2=default

          float temp;

          // --- Primary: Home Assistant sensor ---
          if (!isnan(id(ha_temperature).state)) {
            temp = id(ha_temperature).state;
            current_source = 0;
          }
          // --- Fallback: SHT30 ---
          else if (!isnan(id(ac_temperature).state)) {
            temp = id(ac_temperature).state;
            current_source = 1;
          }
          // --- Last resort default ---
          else {
            temp = 35.0;
            current_source = 2;
          }

          // --- Change detection: print log + update text sensor ---
          if (current_source != last_source) {
            if (current_source == 0) {
              ESP_LOGI("temp_source", "Using Home Assistant sensor");
              id(temp_source).publish_state("homeassistant");
            } else if (current_source == 1) {
              ESP_LOGW("temp_source", "Using SHT30 fallback sensor");
              id(temp_source).publish_state("sht30");
            } else {
              ESP_LOGE("temp_source", "Using DEFAULT temperature (25C)");
              id(temp_source).publish_state("default");
            }
            last_source = current_source;
          }

          // ----------------- Lookup Table -----------------
          float level = 0.0;
          const int N = 10;
          float temps[N]  = {0, 10, 15, 20, 22, 25, 28, 30, 35, 40};
          float levels[N] = {0.818, 0.732, 0.688, 0.643, 0.626, 0.599, 0.571, 0.553, 0.507, 0.463};

          // Linear interpolation
          if (temp <= temps[0]) {
            level = levels[0];
          } else if (temp >= temps[N-1]) {
            level = levels[N-1];
          } else {
            for (int i=0;i<N-1;i++){
              if (temp >= temps[i] && temp <= temps[i+1]){
                float frac = (temp - temps[i]) / (temps[i+1]-temps[i]);
                level = levels[i] + frac*(levels[i+1]-levels[i]);
                break;
              }
            }
          }

          // Clamp output
          if (level < 0.0) level = 0.0;
          if (level > 1.0) level = 1.0;

          id(dac_output).set_level(level);




# number:
#   - platform: template
#     name: "DAC Voltage Level"
#     id: dac_voltage
#     min_value: 0
#     max_value: 1
#     step: 0.001
#     optimistic: true
#     set_action:
#       - output.set_level:
#           id: dac_output
#           level: !lambda "return x;"



# Specify pins used by the board to comunicate with RS485 board - in this case M5STACK
uart:
  tx_pin: GPIO19   # Use GPIO26 for ATOM Tail485 
  rx_pin: GPIO22   # Use GPIO32 for ATOM Tail485
  # For older devices it could also be 2400. But you should start with 9600.
  baud_rate: 9600
  parity: EVEN
#  debug:
#    direction: BOTH
#    dummy_receiver: false
#    after:
#      delimiter: "\n"
#    sequence:
#      - lambda: UARTDebug::log_hex(direction, bytes, ' ');
## SAMSUNG AC CONFIGURATION

# Import custom component from GitHub
external_components:
  - source: github://omerfaruk-aran/esphome_samsung_hvac_bus@main
    components: [samsung_ac]
    refresh: always

button:
  - platform: restart
    name: "Restart"
    entity_category: config


# Configuration of AC component
samsung_ac:

  # For NonNASA devices the following option can be enabled to prevent the device from sleeping when idle. This allows
  # values like internal and external temperature to continue to be tracked when the device isn't in use.
  non_nasa_keepalive: true
  non_nasa_tx_delay_ms: 15   # default 0
  
  # When enabled (set to true), this option will log the messages associated with undefined codes on the device. This is useful for debugging and identifying any unexpected or unknown codes that the device may receive during operation.
  debug_log_undefined_messages: false
  
  # When enabled (set to true), this option logs messages associated with defined codes on the device. This helps in monitoring the behavior of the device by recording the activity related to known, expected codes.
  debug_log_messages: false

  # For half duplex RS485 board with DI, DE, RE, and RO pins, you have to control read/write with control pin
  # see https://microcontrollerslab.com/rs485-serial-communication-esp32-esp8266-tutorial/ for wiring
  #flow_control_pin: GPIO4

  # Capabilities configure the features that all devices of your AC system have (all parts of this section are optional). 
  # All capabilities are off by default, you need to enable only those your devices have.
  # You can override or configure them also on a per-device basis (look below for that).
  capabilities: 
    vertical_swing: true
    horizontal_swing: true
    # Presets define special AC modes like Windfree, Eco, and so on. 
    # The following modes are available: sleep, quiet, fast, longreach, windfree, eco.
    presets: 
      # Short version - a quick and simple way to enable presets.
      quiet: true
      eco: true
      longreach: true
      fast: true

  devices:
    # Skip everything below on the first run! Wait a minute, watch your ESPHome logs until you see the "Discovered devices:" section and you see some addresses:
    # For NASA devices it looks like this (with 1 outdoor device and 4 indoor devices): 
    #   Outdoor: 10.00.00
    #   Indoor:  20.00.00, 20.00.01, 20.00.02, 20.00.03
    # For NonNASA devices it looks like this (with 2 indoor devices): 
    #   Indoor:  00, 01
    # You need these addresses for the the sections below. If you see only a "-" then you should visit the troubleshooting section in the readme. 

    # Repeat everything below for each indoor device address you see in your logs
    - address: "00" # Indoor device address 
      # Each property below is optional (climate, room_temperature etc.) - you can delete those which you dont need.
      # For the names we suggest to choose a combination of room name and the thing it controls. 
      # Configures/overrides the capabilites for this devices. 
      # Look above for all options.
      capabilities: 
        vertical_swing: true
        horizontal_swing: true
        # Presets define special AC modes like Windfree, Eco, and so on. 
        # The following modes are available: sleep, quiet, fast, longreach, windfree, eco.
        presets: 
          # Short version - a quick and simple way to enable presets.
          quiet: true
          eco: true
          # Long version - allows customization, including localized names and control over enabling the presets.
          fast:
            name: "Fast cooling"
            enabled: true

      # Creates climate control in Home Assistant. A climate control combines multiple of the controls below (like temperature, mode etc.)
      climate:
        name: "Bedroom AC"

      # The controls directly below are all included in the climate control. Its adviced to only add the climate control and skip the extra controls.
      room_temperature:
        name: "Bedroom AC temperature"
      target_temperature:
        name: "Bedroom AC target temperature"
      power:
        name: "Bedroom AC power"
      mode:
        name: "Bedroom AC mode"

      #Add automatic_cleaning mode for Samsung AC NASA
            #- Added the "automatic_cleaning" feature to the Samsung AC integration.
            #- This mode ensures hygienic drying of all moisture in the indoor unit after the cooling process is completed.
      automatic_cleaning:
        name: "Bedroom AC automatic clean"

      # If your AC sits near or inside the ceiling, the reported room temperature is often a little bit heigher then whats 
      # measured below. This property can be used to correct that value.
      #room_temperature_offset: -1.4

      # Only supported on NASA devices
      #room_humidity:
      #  name: "Bedroom Samsung humidity"

      # EVA (Evaporator) temperatures - measures refrigerant temperatures at the indoor unit
      # These sensors help monitor system performance and diagnose issues
      # Supported on both NASA and Non-NASA devices
      #indoor_eva_in_temperature:
      #  name: "Bedrrom AC EVA In temperature"
      #indoor_eva_out_temperature:
      #  name: "Bedroom AC EVA Out temperature"
      
    - address: "c8" # Outdoor device address as the following components are dependent on an outdoor unit.
      # This sensor captures and monitors specific error codes returned by the HVAC system.
      # When an error occurs, the sensor detects the error code and updates its value accordingly.
      # Additionally, by using the blueprint available at https://github.com/omerfaruk-aran/esphome_samsung_ac_blueprint,
      # you can automatically send detailed error messages to your mobile devices based on the captured error codes.
      error_code:
        name: error_code
      # This sensor measures the instantaneous power consumption of the outdoor unit in Watts.
      # The captured value represents the current power draw of the outdoor HVAC components, helping track energy usage patterns.
      outdoor_instantaneous_power:
        name: "Outdoor Instantaneous Power"

      # This sensor records the cumulative energy consumption of the outdoor unit in kWh.
      # It calculates the total energy consumed over time, allowing users to monitor and analyze energy efficiency.
      outdoor_cumulative_energy:
        name: "Outdoor Cumulative Energy"

      # This sensor measures the current drawn by the outdoor unit in Amperes.
      # Monitoring current values helps identify electrical irregularities and ensure safe power levels in the system.
      outdoor_current:
        name: "Outdoor Current"

      # This sensor tracks the voltage supplied to the outdoor unit in Volts.
      # Consistent voltage readings indicate stable power delivery, while deviations can help detect electrical issues in the system.
      outdoor_voltage:
        name: "Outdoor Voltage"

      # Only supported on NASA based heatpumps
#      water_temperature:
#        name: "Warm water"
#      water_target_temperature:
#        name: "Hot Water Target Temperature"
#      water_heater_mode:
#        name: "Hotwater Mode"
      outdoor_temperature: # Should be used with outdoor device address
        name: "Outdoor temperature"
```

### Additional Details

_No response_

Also getting the above mentioned error:

"Compiling .pioenvs/samsung-ac/src/esphome/components/samsung_ac/debug_mqtt.cpp.o
Compiling .pioenvs/samsung-ac/src/esphome/components/samsung_ac/log.cpp.o
src/esphome/components/samsung_ac/debug_mqtt.cpp:8:10: fatal error: mqtt_client.h: No such file or directory

***************************************************************************
* Looking for mqtt_client.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:mqtt_client.h"
* Web  > https://registry.platformio.org/search?q=header:%1B%5Bm%1B%5BKmqtt_client.h
*
***************************************************************************

    8 | #include <mqtt_client.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
*** [.pioenvs/samsung-ac/src/esphome/components/samsung_ac/debug_mqtt.cpp.o] Error 1
========================= [FAILED]"

--- a/components/samsung_ac/__init__.py
+++ b/components/samsung_ac/__init__.py
@@ -1,6 +1,14 @@
 import esphome.codegen as cg
 import esphome.config_validation as cv
-from esphome.components import uart, sensor, switch, select, number, climate, text_sensor
+from esphome.components import (
+    uart,
+    sensor,
+    switch,
+    select,
+    number,
+    climate,
+    text_sensor,
+)
 from esphome.const import (
     CONF_ID,
     DEVICE_CLASS_TEMPERATURE,
@@ -169,11 +177,15 @@ def custom_sensor_schema(
         accuracy_decimals=accuracy_decimals,
         device_class=device_class,
         state_class=state_class,
-        entity_category=entity_category
-    ).extend({
-        cv.Optional(CONF_DEVICE_CUSTOM_MESSAGE, default=message): cv.hex_int,
-        cv.Optional(CONF_DEVICE_CUSTOM_RAW_FILTERS, default=raw_filters): sensor.validate_filters,
-    })
+        entity_category=entity_category,
+    ).extend(
+        {
+            cv.Optional(CONF_DEVICE_CUSTOM_MESSAGE, default=message): cv.hex_int,
+            cv.Optional(
+                CONF_DEVICE_CUSTOM_RAW_FILTERS, default=raw_filters
+            ): sensor.validate_filters,
+        }
+    )
 
     return schema
 
@@ -207,6 +219,7 @@ def error_code_sensor_schema(message: int):
         icon="mdi:alert",
         entity_category="diagnostic",
     )
+
 
 DEVICE_SCHEMA = cv.Schema(
     {
@@ -243,12 +256,10 @@ DEVICE_SCHEMA = cv.Schema(
         cv.Optional(CONF_DEVICE_WATER_OUTLET_TARGET): NUMBER_SCHEMA,
         cv.Optional(CONF_DEVICE_WATER_TARGET_TEMPERATURE): NUMBER_SCHEMA,
         cv.Optional(CONF_DEVICE_POWER): switch.switch_schema(
-            Samsung_AC_Switch,
-            icon="mdi:power"
+            Samsung_AC_Switch, icon="mdi:power"
         ),
         cv.Optional(CONF_DEVICE_AUTOMATIC_CLEANING): switch.switch_schema(
-            Samsung_AC_Switch,
-            icon="mdi:broom"
+            Samsung_AC_Switch, icon="mdi:broom"
         ),
         cv.Optional(CONF_DEVICE_WATER_HEATER_POWER): switch.switch_schema(
             Samsung_AC_Switch
@@ -312,15 +323,18 @@ DEVICE_SCHEMA = cv.Schema(
                 cv.Optional(CONF_DEVICE_CUSTOM_MESSAGE, default=0x24FC): cv.hex_int,
             }
         ),
-        cv.Optional(CONF_DEVICE_OUT_OPERATION_ODU_MODE_TEXT): text_sensor.text_sensor_schema(
+        cv.Optional(
+            CONF_DEVICE_OUT_OPERATION_ODU_MODE_TEXT
+        ): text_sensor.text_sensor_schema(
             icon="mdi:fan",
             entity_category="diagnostic",
         ),
-        cv.Optional(CONF_DEVICE_OUT_OPERATION_HEATCOOL_TEXT): text_sensor.text_sensor_schema(
+        cv.Optional(
+            CONF_DEVICE_OUT_OPERATION_HEATCOOL_TEXT
+        ): text_sensor.text_sensor_schema(
             icon="mdi:thermometer",
             entity_category="diagnostic",
         ),
-
     }
 )
 
@@ -357,7 +371,9 @@ CONFIG_SCHEMA = (
             cv.Optional(CONF_DEBUG_LOG_MESSAGES, default=False): cv.boolean,
             cv.Optional(CONF_DEBUG_LOG_MESSAGES_RAW, default=False): cv.boolean,
             cv.Optional(CONF_NON_NASA_KEEPALIVE, default=False): cv.boolean,
-            cv.Optional(CONF_NON_NASA_TX_DELAY_MS, default=0): cv.int_range(min=0, max=1000),
+            cv.Optional(CONF_NON_NASA_TX_DELAY_MS, default=0): cv.int_range(
+                min=0, max=1000
+            ),
             cv.Optional(CONF_DEBUG_LOG_UNDEFINED_MESSAGES, default=False): cv.boolean,
             cv.Optional(CONF_CAPABILITIES): CAPABILITIES_SCHEMA,
             cv.Optional(CONF_DEBUG_LOG_MESSAGES_ON_CHANGE, default=False): cv.boolean,
@@ -387,9 +403,11 @@ async def to_code(config):
         # setup capabilities
         capabilities = device.get(CONF_CAPABILITIES, config.get(CONF_CAPABILITIES, {}))
 
-        cg.add(var_dev.set_supports_fan_modes(
-            capabilities.get(CONF_CAPABILITIES_FAN_MODES, True)
-        ))
+        cg.add(
+            var_dev.set_supports_fan_modes(
+                capabilities.get(CONF_CAPABILITIES_FAN_MODES, True)
+            )
+        )
 
         if CONF_CAPABILITIES_VERTICAL_SWING in capabilities:
             cg.add(
@@ -430,12 +448,8 @@ async def to_code(config):
 
                 cg.add(
                     var_dev.add_alt_mode(
-                        preset_conf.get(
-                            CONF_PRESET_NAME, preset_info["displayName"]
-                        ),
-                        preset_conf.get(
-                            CONF_PRESET_VALUE, preset_info["value"]
-                        ),
+                        preset_conf.get(CONF_PRESET_NAME, preset_info["displayName"]),
+                        preset_conf.get(CONF_PRESET_VALUE, preset_info["value"]),
                     )
                 )
 
@@ -500,7 +514,6 @@ async def to_code(config):
                 text_sensor.new_text_sensor,
                 var_dev.set_outdoor_operation_heatcool_text_sensor,
             ),
-
         }
 
         # Iterate over the actions
@@ -563,8 +576,12 @@ async def to_code(config):
             cg.add(var_dev.set_climate(var_cli))
 
             # Optional UI mapping: expose Samsung Auto as HA Heat/Cool
-            cg.add(var_dev.set_map_auto_to_heat_cool(device.get(CONF_MAP_AUTO_TO_HEAT_COOL, False)))
-    
+            cg.add(
+                var_dev.set_map_auto_to_heat_cool(
+                    device.get(CONF_MAP_AUTO_TO_HEAT_COOL, False)
+                )
+            )
+
         if CONF_DEVICE_CUSTOM in device:
             for cust_sens in device[CONF_DEVICE_CUSTOM]:
                 sens = await sensor.new_sensor(cust_sens)
@@ -590,6 +607,24 @@ async def to_code(config):
                 )
 
         cg.add(var.register_device(var_dev))
+
+    # If debug MQTT is enabled on ESP32, we need ESP-IDF's built-in mqtt component.
+    # ESPHome may exclude it by default to reduce compile time, so re-enable it only when needed.
+    if CORE.is_esp32 and config[CONF_DEBUG_MQTT_HOST]:
+        cg.add_define("SAMSUNG_AC_DEBUG_MQTT")
+        try:
+            from esphome.components.esp32 import include_builtin_idf_component
+
+            include_builtin_idf_component("mqtt")
+        except Exception:
+            try:
+                from esphome.components.esp32 import include_idf_component
+
+                include_idf_component("mqtt")
+            except Exception:
+                from esphome.components.esp32 import add_idf_component
+
+                add_idf_component("mqtt")
 
     cg.add(
         var.set_debug_mqtt(

--- a/components/samsung_ac/debug_mqtt.cpp
+++ b/components/samsung_ac/debug_mqtt.cpp
@@ -3,94 +3,102 @@
 
 #if defined(USE_ESP8266)
 #include <AsyncMqttClient.h>
-AsyncMqttClient *mqtt_client{nullptr};
-#elif defined(USE_ESP32)
+static AsyncMqttClient *mqtt_client{nullptr};
+#endif
+
+#if defined(USE_ESP32) && defined(SAMSUNG_AC_DEBUG_MQTT)
 #include <mqtt_client.h>
-esp_mqtt_client_handle_t mqtt_client{nullptr};
+static esp_mqtt_client_handle_t mqtt_client{nullptr};
 #endif
 
 namespace esphome
 {
-    namespace samsung_ac
+namespace samsung_ac
+{
+    bool debug_mqtt_connected()
     {
-        bool debug_mqtt_connected()
-        {
-#if defined(USE_ESP8266) || defined(USE_ESP32)
-            if (mqtt_client == nullptr)
-                return false;
-
 #if defined(USE_ESP8266)
-            return mqtt_client->connected();
-#elif defined(USE_ESP32)
-            return true;
-#endif
+        if (mqtt_client == nullptr)
+            return false;
+        return mqtt_client->connected();
+
+#elif defined(USE_ESP32) && defined(SAMSUNG_AC_DEBUG_MQTT)
+        if (mqtt_client == nullptr)
+            return false;
+
+        return true;
+
 #else
         return false;
 #endif
-        }
+    }
 
-        void debug_mqtt_connect(const std::string &host, const uint16_t port, const std::string &username, const std::string &password)
-        {
-            if (host.empty())
-                return;
+    void debug_mqtt_connect(const std::string &host, const uint16_t port,
+                            const std::string &username, const std::string &password)
+    {
+        if (host.empty())
+            return;
 
 #if defined(USE_ESP8266)
-            if (mqtt_client == nullptr)
-            {
-                mqtt_client = new AsyncMqttClient();
-                mqtt_client->setServer(host.c_str(), port);
-                if (!username.empty())
-                    mqtt_client->setCredentials(username.c_str(), password.c_str());
-            }
+        if (mqtt_client == nullptr)
+        {
+            mqtt_client = new AsyncMqttClient();
+            mqtt_client->setServer(host.c_str(), port);
+            if (!username.empty())
+                mqtt_client->setCredentials(username.c_str(), password.c_str());
+        }
 
-            if (!mqtt_client->connected())
-                mqtt_client->connect();
+        if (!mqtt_client->connected())
+            mqtt_client->connect();
 
-#elif defined(USE_ESP32)
-            if (mqtt_client == nullptr)
-            {
+#elif defined(USE_ESP32) && defined(SAMSUNG_AC_DEBUG_MQTT)
+        if (mqtt_client == nullptr)
+        {
 #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 0, 0)
-                // For ESP-IDF v5.0 and above
-                std::string uri = "mqtt://" + host + ":" + std::to_string(port);
-                esp_mqtt_client_config_t mqtt_cfg = {};
-                mqtt_cfg.broker.address.uri = uri.c_str();
-                if (!username.empty())
-                {
-                    mqtt_cfg.credentials.username = username.c_str();
-                    mqtt_cfg.credentials.authentication.password = password.c_str();
-                }
-#else
-                // For ESP-IDF versions below v5.0
-                esp_mqtt_client_config_t mqtt_cfg = {};
-                mqtt_cfg.host = host.c_str();
-                mqtt_cfg.port = port;
-                if (!username.empty())
-                {
-                    mqtt_cfg.username = username.c_str();
-                    mqtt_cfg.password = password.c_str();
-                }
-#endif
-                mqtt_client = esp_mqtt_client_init(&mqtt_cfg);
-                esp_mqtt_client_start(mqtt_client);
+            std::string uri = "mqtt://" + host + ":" + std::to_string(port);
+            esp_mqtt_client_config_t mqtt_cfg = {};
+            mqtt_cfg.broker.address.uri = uri.c_str();
+            if (!username.empty())
+            {
+                mqtt_cfg.credentials.username = username.c_str();
+                mqtt_cfg.credentials.authentication.password = password.c_str();
             }
-
+#else
+            esp_mqtt_client_config_t mqtt_cfg = {};
+            mqtt_cfg.host = host.c_str();
+            mqtt_cfg.port = port;
+            if (!username.empty())
+            {
+                mqtt_cfg.username = username.c_str();
+                mqtt_cfg.password = password.c_str();
+            }
 #endif
+            mqtt_client = esp_mqtt_client_init(&mqtt_cfg);
+            esp_mqtt_client_start(mqtt_client);
         }
-
-        bool debug_mqtt_publish(const std::string &topic, const std::string &payload)
-        {
-#if defined(USE_ESP8266) || defined(USE_ESP32)
-            if (mqtt_client == nullptr)
-                return false;
-
-#if defined(USE_ESP8266)
-            return mqtt_client->publish(topic.c_str(), 0, false, payload.c_str()) != 0;
-#elif defined(USE_ESP32)
-            return esp_mqtt_client_publish(mqtt_client, topic.c_str(), payload.c_str(), payload.length(), 0, false) != -1;
+#else
+        (void)port;
+        (void)username;
+        (void)password;
 #endif
+    }
+
+    bool debug_mqtt_publish(const std::string &topic, const std::string &payload)
+    {
+#if defined(USE_ESP8266)
+        if (mqtt_client == nullptr)
+            return false;
+        return mqtt_client->publish(topic.c_str(), 0, false, payload.c_str()) != 0;
+
+#elif defined(USE_ESP32) && defined(SAMSUNG_AC_DEBUG_MQTT)
+        if (mqtt_client == nullptr)
+            return false;
+        return esp_mqtt_client_publish(mqtt_client, topic.c_str(), payload.c_str(),
+                                       payload.length(), 0, false) != -1;
 #else
         return false;
 #endif
-        }
-    } // namespace samsung_ac
+    }
+
+} // namespace samsung_ac
 } // namespace esphome

